requests/adapters.py

```diff
@@ -201,8 +260,10 @@ def proxy_manager_for(self, proxy, **proxy_kwargs):
         return manager
 
     def cert_verify(self, conn, url, verify, cert):
```

Comment:
I think you'll want to document the method as deprecated, but leave it as-is. Yes, it doesn't work the way most people want, but leaving the signature there with a behavior change is an API breakage, too.

What seems less invasive to me is to instead is to call this method in the default implementation conditionally based on a feature flag which defaults to `False`. That way existing users see no behavior change.

Reply:
I see, I'll revert to the original the body of `cert_verify`. Do we want a warning (like `urllib3.connection` does) notifying that the method has been deprecated if it's called, or just a comment in the docstring.

If I understand your suggestion for a feature flag, the constructor for `Session` objects would take an optional boolean that defaults to False. When True, it would cause to skip the call to `cert_verify` from within `HTTPAdapter.send`. Given that the tests pass without the feature flag (and with the original `cert_verify` code), are there advantages to having this feature-flag? One advantage I can see in the case of having a warning produced by `cert_verify` being called, is that it could get the attention of developers to switch to using the feature-flag. But then, wouldn't we be stuck with having that feature-flag? I haven't written software that's as widely deployed as requests, so I don't have lots of experience about which types of changes are preferable when fixing a bug, and what are the consequences downstream. For example, I would have though it preferable for code to break than to make an unintended unverified request.

Reply:
Well, if you're not breaking any APIs no flag is necessary, but I'm a bit surprised it's not breaking something. I can't say, though, without thoroughly re-acquainting myself with urllib3 and requests code. 

Sorry I can't be more help, but I'm no longer using either one in any particularly complex situations so it's hard to gather the energy to briefly dive back into the codebases for a review. Maybe in the next couple weeks I'll feel inspired to do some digging, but I don't know.

Reply:
So what was happening before this change was that in `adapter.HTTPAdaper.send` we were first calling `get_connection`, then calling `cert_verify`. The `get_connection` call was getting to the right `PoolManager` and returning a `ConnectionPool` and then the `cert_verify` was (illegally) modifying that `ConnectionPool` object. It appears that the first time you call `cert_verify` on a connection pool, it behaves like a call to `HTTPSConnectionPool.set_cert`. According to the docs "this method should only be called once, before the connection is used". 

The reason I think this change is working without breaking anything is that now `get_connection` uses `verify` and `cert` to call `_build_conn_kwargs` which is passed to the `PoolManager` to return the right `ConnectionPool`. Keep in mind that `_build_conn_kwargs` uses the same logic as `cert_verify` except it builds a kwarg dict instead of modifying a `ConnectionPool` in place. So, by using these connection kwargs, the `PoolManager` is returning to us the "correct" `ConnectionPool` and subsequently, when we call `cert_verify`, it sets some attributes on the connection, but those attributes should already exist and have the correct values (as per the `conn_kwargs`). So, in effect `cert_verify` should be a no-op because we used the same values in `conn_kwargs` when we got the `ConnectionPool` from the `PoolManager`.

Did this make sense? I'm trying to describe at a high-ish level what I think is happening. Like you, I am not intimately familiar with the inner workings of urllib3, but these were my conclusions from running the failing test I added through `pdb` and looking at what we got back from urllib3.

Reply:
Also, I understand you're not a maintainer here, and understanding why this works would require some non-trivial digging. Do you think I should ping the maintainers ?